### PR TITLE
doc: Clarify that nix-shell still uses shell from host environment

### PR DESCRIPTION
**Pull request by CyberShadow**
# Motivation

This came as a surprise to me, and even other experienced users from the Matrix channel.

# Context

This is a documentation improvement which adds clarity to the semantics of the `nix-shell` command.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by CyberShadow** [comment]
Sorry, this is not accurate. The host shell is used only for `shell.nix`; will clarify and reopen.

**Comment by CyberShadow** [comment]
> will clarify and reopen

I think the source of my confusion is just that `shell.nix` is not the right tool for when you want to fully encapsulate your dependencies in a single file. Unless I'm missing something, `nix-shell` will still use the search path from its environment to pick the shell to run in the environment created by a `shell.nix` file, and now I'm not sure if this is something that can/should be expressed in the documentation.

**Comment by CyberShadow** [comment]
Demonstration:

```console
$ cat > shell.nix <<EOF
#!/usr/bin/env -S -u NIX_PATH nix-shell --pure
let
  pkgs = import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/854fdc68881791812eddd33b2fed94b954979a8e.tar.gz") {};
in
pkgs.mkShell {
  buildInputs = with pkgs; [
    bash
  ];
}
EOF
$ nix-shell
error:
       … while calling the 'import' builtin

         at «string»:1:2:

            1| (import <nixpkgs> {}).bashInteractive
             |  ^

       … while calling the 'findFile' builtin

         at «string»:1:9:

            1| (import <nixpkgs> {}).bashInteractive
             |         ^

       error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

       at «none»:0: (source not available)
will use bash from your environment

[nix-shell:~/tmp]$ readlink /proc/$$/exe
/usr/bin/bash
```

**Comment by CyberShadow** [comment]
Revisiting this, I still think that at least a small clarification would help.

My confusion was that I expected the shell (Bash) executable to be obtained from the same `pkgs` as the one used to call `mkShell`, but the logic is actually different (explained lower in the "Environment variables" section), which can lead `nix-shell` to use one `nixpkgs` version for the shell executable, and another for `mkShell`.

File: doc/manual/src/command-ref/nix-shell.md [hunk 1]
```diff
@@ -88,7 +88,9 @@ All options not listed here are passed to `nix-store
   cleared before the interactive shell is started, so you get an
   environment that more closely corresponds to the “real” Nix build. A
   few variables, in particular `HOME`, `USER` and `DISPLAY`, are
-  retained.
+  retained.  Note that the shell used to run commands is obtained from
+  `NIX_BUILD_SHELL` / `<nixpkgs>` from `NIX_PATH`, and therefore not
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-shell.md:92]: Apparently [the note on `NIX_BUILD_SHELL`](https://github.com/NixOS/nix/pull/8809/files#diff-3eb54def6708c86fc4695323b8a91cd918c8e7d63b80d14d26a0385952c6176bL119) is wrong. This is what actually happens:

https://github.com/NixOS/nix/blob/621c23bbea9db9138d64a8111f6183672d85c5af/src/nix-build/nix-build.cc#L432-L443

**Review by fricklerhandwerk — COMMENTED** [review]
> Demonstration:

Aghr, `nix-shell` is the worst, and just about everything involved in it, inlcuding `pkgs.mkShell`. The number of issues this keeps spawning is staggering. I suggest we document all of that in excruciating detail, and re-think the entire complex from scratch: https://github.com/NixOS/nix.dev/issues/135#issuecomment-2233802220 https://github.com/NixOS/nix/pull/4702#issuecomment-2233787312 

@CyberShadow could you please fix up the details and also add your example?  Then I'll merge quickly. I suggest putting the full explanation under `NIX_BUILD_SHELL` and linking there from `--pure`. Please check other files for how to make anchors, and test the resulting manual *and* `man` and `--help` pages that it renders correctly.

**Comment by CyberShadow** [comment]
I tried to add the example but it doesn't look very good in `man`:

```

Environment variables
       •  NIX_BUILD_SHELL

          Shell used to start the interactive environment. Defaults to the bash from
          bashInteractive found in <nixpkgs>, falling back to the bash found in PATH
          if not found.

          Note  that  the  default  shell  obtained  using  the method above may not
          necessarily be the same as any shells  requested  in  path.  For  example,
          consider:

       #!/usr/bin/env -S nix-shell --pure
       let
       pkgs = import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/854fdc6888
1791812eddd33b2fed94b954979a8e.tar.gz") {};
       in
       pkgs.mkShell {
       buildInputs = pkgs.bashInteractive;
       }

              Despite  --pure,  the  above  will  not result in a fully reproducible
              shell environment.

Common Environment Variables
```

File: doc/manual/src/command-ref/nix-shell.md (hunk 1 above)
Comment (CyberShadow) [reply doc/manual/src/command-ref/nix-shell.md]: I went over the code and the only discrepancy I've noticed is the package used -  `bash` vs. `bashInteractive`, is there anything else I missed?

File: doc/manual/src/command-ref/nix-shell.md [hunk 2]
```diff
@@ -112,11 +114,28 @@ All options not listed here are passed to `nix-store
...
+- <span id="env-NIX_BUILD_SHELL">[`NIX_BUILD_SHELL`](#env-NIX_BUILD_SHELL)</span>
 
   Shell used to start the interactive environment. Defaults to the
-  `bash` found in `<nixpkgs>`, falling back to the `bash` found in
-  `PATH` if not found.
+  `bash` from `bashInteractive` found in `<nixpkgs>`, falling back to
+  the `bash` found in `PATH` if not found.
+
+  Note that the default shell obtained using the method above may not
+  necessarily be the same as any shells requested in *path*. For
```
Comment (fricklerhandwerk) [inline doc/manual/src/command-ref/nix-shell.md:124]: What's `*path*` here, and what do you mean by request?

File: doc/manual/src/command-ref/nix-shell.md (hunk 1 above)
Comment (fricklerhandwerk) [reply doc/manual/src/command-ref/nix-shell.md]: It seems I had some blooper there. Not sure what my problem was... Looks fine now.

**Review by fricklerhandwerk — COMMENTED** [review]
So much for "merge quickly"... Please excuse the long delay, other things took priority. I need a bit of clarification on what exactly you want to say, otherwise this should be good to go.

I'd say don't worry about the rendering. lowdown is janky and all of this should work differently anyway; this would another piece of motivation to change it.

File: doc/manual/src/command-ref/nix-shell.md (hunk 2 above)
Comment (CyberShadow) [reply doc/manual/src/command-ref/nix-shell.md:124]: > What's `*path*` here

It's the CLI argument:

https://github.com/CyberShadow/nix/blob/pull-20230811-081846/doc/manual/src/command-ref/nix-shell.md?plain=1#L16

It is pretty far away from there, maybe there is a way to refer to it less ambiguously?

> and what do you mean by request?

Indeed not sure if it is the best word bug e.g. if *path* is a file with an expression that calls `mkShell` or anything else with a specific `bash` in e.g. `buildInputs`, the goal is to communicate that that `bash` is not the `bash` that will be used by `nix-shell`.

**Comment by fricklerhandwerk** [comment]
Thanks @CyberShadow for your patience. Sometimes things appear in my queue at an unfortunate time and then slip through the cracks.

**Pull request merged**